temporal-sdk/src/test/java/io/temporal/workflow/LocalAsyncCompletionWorkflowTest.java

```diff
@@ -0,0 +1,145 @@
+      TestWorkflowRule.newBuilder()
+          .setWorkerOptions(
+              WorkerOptions.newBuilder()
+                  .setMaxConcurrentActivityExecutionSize(MAX_CONCURRENT_ACTIVITIES)
+                  .setActivityPollThreadCount(5)
+                  .build())
+          .setWorkflowTypes(TestWorkflowImpl.class)
+          .setActivityImplementations(new AsyncActivityWithManualCompletion())
+          .setUseExternalService(Boolean.parseBoolean(System.getenv("USE_DOCKER_SERVICE")))
+          .setTarget(System.getenv("TEMPORAL_SERVICE_ADDRESS"))
+          .build();
+
+  private static final AtomicInteger concurrentActivitiesCount = new AtomicInteger(0);
+
+  @WorkflowInterface
+  public interface TestWorkflow {
+
+    @WorkflowMethod
+    String execute(String taskQueue);
+  }
+
+  public static class TestWorkflowImpl implements TestWorkflow {
+
+    @Override
+    public String execute(String taskQueue) {
+      TestActivity activity =
+          Workflow.newActivityStub(
+              TestActivity.class,
+              ActivityOptions.newBuilder()
+                  .setScheduleToStartTimeout(Duration.ofSeconds(10))
+                  .setScheduleToCloseTimeout(Duration.ofSeconds(10))
+                  .setHeartbeatTimeout(Duration.ofSeconds(1))
+                  .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(1).build())
+                  .build());
+      List<Promise<Integer>> promises = new ArrayList<>();
+      for (int i = 0; i < 10; i++) {
+        promises.add(Async.function(() -> activity.execute(2)));
+      }
+      Promise.allOf(promises).get();
+      for (Promise<Integer> promise : promises) {
+        if (promise.getFailure() != null) {
+          return "exception";
+        }
+        if (promise.get() != 4) { // All activities compute 2 * 2
+          return "wrong result";
+        }
+      }
+      return "success";
+    }
+  }
+
+  @ActivityInterface
+  public interface TestActivity {
+
+    @ActivityMethod
+    int execute(int value);
+  }
+
+  public static class AsyncActivityWithManualCompletion implements TestActivity {
+
+    @Override
+    public int execute(int value) {
+      int concurrentActivities = concurrentActivitiesCount.incrementAndGet();
+      if (concurrentActivities > MAX_CONCURRENT_ACTIVITIES) {
+        throw new RuntimeException(
+            String.format(
+                "More than %d activities was running concurrently!", MAX_CONCURRENT_ACTIVITIES));
+      }
+      ActivityExecutionContext context = Activity.getExecutionContext();
+      context.heartbeat(value);
+      ActivityCompletionClient completionClient = context.useLocalManualCompletion();
+      try {
+        Thread.sleep(500);
+        completionClient.complete(context.getTaskToken(), value * 2);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        e.printStackTrace();
+        completionClient.completeExceptionally(context.getTaskToken(), e);
```

Comment:
I would move completion to a separate thread just to increase code coverage. See [HelloAsyncActivityCompletion](https://github.com/temporalio/samples-java/blob/8218f4114e52417f8d04175b67027ff0af4fb73c/src/main/java/io/temporal/samples/hello/HelloAsyncActivityCompletion.java#L99).

Reply:
Done